### ISSUE TITLE
504 Gateway Timeout Error - when the cluster have a lot of nodes

### ISSUE THREAD

**Issue by sam-som** [opencost/opencost#1841]
**Describe the bug**
When a cluster have a lot of nodes (200+ in our case), we would get a `504 Gateway Timeout Error`
<img width="524" alt="Screenshot 2023-04-04 at 8 29 33 PM" src="https://user-images.githubusercontent.com/37458846/229982807-8b52d67f-1ca4-442a-9dee-fa1ab6e68d5f.png">

**To Reproduce**

1. Deploy `opencost` on a cluster with a lot of nodes (200+ for example)
2. Hit the API below (or load the UI)
```shell
curl -X GET \
    'https://{opencost-host}/model/allocation/compute?window=6d&aggregate=namespace&step=1d&accumulate=false'
```

**Expected behavior**
The data is returned by the API

**Actual behavior**
```shell
<html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
</body>
</html>
```

**Which version of OpenCost are you using?**
Opencost app version `1.101.2`
Helm version `1.10.0`

**Additional context**
I tried to set the log level to `trace` and see the following error
```shell
2023-04-05T04:27:49.860564281Z ERR CostModel.ComputeAllocation: query context error Errors:
  Request Error: Prometheus communication error: 503 (Service Unavailable) URL: 'http://prometheus.kube-system.svc:9090/api/v1/query?query=sum%28increase%28container_network_transmit_bytes_total%7Bpod%21%3D%22%22%2C+container%3D%22POD%22%7D%5B1d%5D%29%29+by+%28pod_name%2C+pod%2C+namespace%2C+cluster_id%29&time=1680666912', Request Headers: 'map[]', Headers: '{ Content-Type: [ application/json ], Date: [ Wed, 05 Apr 2023 04:27:45 GMT ] }', Body: '{"status":"error","errorType":"timeout","error":"query timed out in expression evaluation"}' Query: 'sum(increase(container_network_transmit_bytes_total{pod!="", container="POD"}[1d])) by (pod_name, pod, namespace, cluster_id)'
  Parse Error: Prometheus communication error: sum(increase(container_network_transmit_bytes_total{pod!="", container="POD"}[1d])) by (pod_name, pod, namespace, cluster_id)
for Query: sum(increase(container_network_transmit_bytes_total{pod!="", container="POD"}[1d])) by (pod_name, pod, namespace, cluster_id)
```

**Comment by sam-som** [comment]
Note:
I have the same setup on other clusters that have less than 70 nodes and those `opencost` instances worked without this issue. 
I'm not very familiar with `prometheus` and hoping someone else has better insight on why the `prometheus` query timeout (and possibly a solution)

**Comment by mattray** [comment]
Could you try your query with less resolution to see if that helps? 
```&resolution=5m```

This sounds very similar to https://github.com/opencost/opencost/issues/1693

**Comment by sam-som** [comment]
Hi @mattray 
I have tried `&resolution=5m` and it improve the performance slightly, but we still cannot get anything larger than `3d` windowed.
Here's a small benchmark that i did
```
time curl -X GET \
    'https://<host>/model/allocation/compute?window=3d&aggregate=namespace&step=1d&accumulate=false'
curl -X GET   0.03s user 0.02s system 0% cpu 1:43.64 total
```

```
time curl -X GET \
    'https://<host>/model/allocation/compute?window=3d&aggregate=namespace&step=1d&accumulate=false&resolution=5m'
curl -X GET   0.03s user 0.02s system 0% cpu 1:41.58 total
```

```
time curl -X GET \
    'https://<host>/model/allocation/compute?window=3d&aggregate=namespace&step=1d&accumulate=false&resolution=30m'
curl -X GET   0.02s user 0.02s system 0% cpu 1:22.36 total
```

**Comment by AjayTripathy** [comment]
Hey @sam-som , our advice here would be to run 1d queries in batch and export the data yourself, add more memory to the backing prometheus, or consider the community edition of our enterprise solution, Kubecost.

**Comment by sam-som** [comment]
@AjayTripathy Is there a guide or document about setting up batch and data export for opencost? I understand that opencost is relatively new, but im wondering if someone already did something similar before. 
Do you have a recommendation for datastore if we want to store 30 days worth of data and display those data in Grafana?

**Comment by mattray** [comment]
There's a new feature coming in the next release (103) for CSV export that might make things easier: https://github.com/opencost/opencost/pull/1792
Docs are inbound too

**Comment by debMan** [comment]
Same issue

**Comment by laurentiuspurba** [comment]
I tried to mimic this https://github.com/kubecost/cost-analyzer-helm-chart/blob/v1.98/cost-analyzer/charts/prometheus/values.yaml#L1208 but the result is the same. I wish I have a magic wand... 😁

**Comment by mattray** [comment]
@laurentiuspurba unfortunately that won't work with OpenCost because we don't manage the Prometheus installation.

**Comment by laurentiuspurba** [comment]
> @laurentiuspurba unfortunately that won't work with OpenCost because we don't manage the Prometheus installation.

@mattray Thanks for your response. Just curious, if there is nothing wrong with Prometheus all pods are up and running, resources are good, but when opening Opencost, I'm getting 504, isn't it the problem that OpenCost could not get data from Prometheus or the way Opencost polling the data could have been an issue.

Let say I have a database up and running, and my web application cannot display the data, is it the problem with database or my application? Again, I just wanted to know, and would like to participate on things that can make Opencost better.

**Comment by mattray** [comment]
@laurentiuspurba we're continuing to make OpenCost-side performance improvements and we'd welcome benchmarking and/or suggestions for Prometheus tuning, but it's not meant to be a longterm database for large amounts of data. The default for Prometheus is to only keep 15 days and there are several tools for exporting the data (CSV, Parquet, API).

**Comment by debMan** [comment]
We resolved the same problem by customizing `ui/default.nginx.conf`. You can refer [here](https://github.com/snapp-incubator/opencost/commit/95f85c5b595ec11cb50b90d4563ac0742287a8ad) for more details.

https://github.com/opencost/opencost/blob/9c1f94b12d5ebba4c619ad7056d600e65fe902c3/ui/default.nginx.conf.template#L69-L71

It would be nice to make it configurable by env variables or something else.

**Comment by laurentiuspurba** [comment]
Hi @debMan thanks for your response. QQ. How did you configure this after you update the nginx configuration? Did you rebuild the OpenCost image and deploy it to your artifact registry such as quay? Really appreciate your help and pointers on this.

**Comment by debMan** [comment]
Yes @laurentiuspurba, we build and push the built image to local artifact repository and use it.

**Comment by mattray** [comment]
I'd be happy to merge PRs for making those timeouts configurable. We'd need environment variables like
```
PROXY_CONNECT_TIMEOUT
PROXY_SEND_TIMEOUT
PROXY_READ_TIMEOUT
```
for the `default.nginx.conf.template` and we could populate those from the Helm chart (another PR)

**Comment by laurentiuspurba** [comment]
> Yes @laurentiuspurba, we build and push the built image to local artifact repository and use it.

Great. Thank you @debMan

**Comment by kastl-ars** [comment]
This is not stale, this is still an issue...

**Comment by Kush172005** [comment]
Hi @mattray,

I took a look into this and it seems like the fix is already implemented in the current codebase.

In the UI, the nginx config already supports configurable timeouts using `PROXY_CONNECT_TIMEOUT`, `PROXY_SEND_TIMEOUT`, and `PROXY_READ_TIMEOUT`, with defaults set to 180s. These can be overridden via environment variables.

On the Helm side as well, under `opencost.ui.nginx` in `values.yaml`, there are options for `proxyConnectTimeout`, `proxySendTimeout`, and `proxyReadTimeout`, which are used when generating the nginx config.

So overall, both approaches (env vars and Helm config) are already supported.

I didn’t find this clearly documented though, so it might help to add a short note for users who run into 504 timeouts on larger clusters.

If this matches the intended behavior, we could close this issue. Happy to open a small PR to add documentation if needed.
